### PR TITLE
Improve performance and precision on movies AI

### PR DESCRIPTION
**Pull request by fjsj**
https://youtu.be/HSA0eleAlbA

### CODE REVIEW

File: example/example/settings.py [hunk 1]
```diff
@@ -180,6 +180,7 @@
 # Example specific settings:
 
 WEATHER_API_KEY = os.getenv("WEATHER_API_KEY")  # get for free at https://www.weatherapi.com/
+JINA_API_KEY = os.getenv("JINA_API_KEY")  # get for free at https://jina.ai/
```
Comment (fjsj) [inline example/example/settings.py:183]: With API key we get larger limits. It's free.

File: example/movies/ai_assistants.py [hunk 2]
```diff
@@ -28,31 +44,38 @@ class IMDbScraper(AIAssistant):
...
+        "Make search queries like:\n"
+        "- IMDb page of <queried movie here>\n"
+        "Then check results, scrape the IMDb URL, process the page, and produce an output like this: \n"
+        "- IMDb URL: ...\n"
+        "- IMDb Rating: ...\n"
+        "- IMDb Page: <Markdown content of the IMDb page>"
     )
     name = "IMDb Scraper"
     model = "gpt-4o-mini"
-    structured_output = IMDbMovie
```
Comment (fjsj) [inline example/movies/ai_assistants.py:39]: structured_output is slower, and not necessary here.

File: example/movies/ai_assistants.py [hunk 3]
```diff
@@ -28,31 +44,38 @@ class IMDbScraper(AIAssistant):
...
+    tool_max_concurrency = 4
```
Comment (fjsj) [inline example/movies/ai_assistants.py:56]: With RateLimitedBraveSearch, we can run tools concurrently and get faster responses. Thanks to this, we can have responses overall from 5 to 30 seconds.

File: example/movies/ai_assistants.py [hunk 4]
```diff
@@ -28,31 +44,38 @@ class IMDbScraper(AIAssistant):
...
-        """Scrape the IMDb URL and return the content as markdown."""
-        return requests.get("https://r.jina.ai/" + url, timeout=20).text[:10000]
+        """Scrape the IMDb URL and return the content as Markdown."""
+        return requests.get(
+            "https://r.jina.ai/" + url,
+            headers={
+                "Authorization": "Bearer " + settings.JINA_API_KEY,
+            },
+            timeout=20,
+        ).text[:30000]
```
Comment (fjsj) [inline example/movies/ai_assistants.py:74]: A bit more text doesn't affect performance.

File: example/movies/ai_assistants.py [hunk 5]
```diff
@@ -63,18 +86,23 @@ class MovieRecommendationAIAssistant(AIAssistant):
     id = "movie_recommendation_assistant"  # noqa: A003
     instructions = (
         "You're a helpful movie recommendation assistant. "
-        "Help the user find movies to watch and manage their movie backlogs. "
-        "Use the provided functions to answer questions and run operations.\n"
+        "Use the provided functions to answer queries and run operations.\n"
+        "Use the search function to find movie recommendations based on user's query.\n"
```
Comment (fjsj) [inline example/movies/ai_assistants.py:90]: Had to add this line to force search. In some cases it was hallucinating URLs. For more popular movies, it gets them right, though.

File: example/movies/ai_assistants.py [hunk 6]
```diff
@@ -93,10 +121,14 @@ def get_instructions(self):
...
         return [
-            BraveSearch.from_api_key(
+            RateLimitedBraveSearch.from_api_key(
                 api_key=settings.BRAVE_SEARCH_API_KEY, search_kwargs={"count": 5}
             ),
-            IMDbScraper().as_tool(description="IMDb Scraper to get the IMDb data a given movie."),
+            IMDbScraper().as_tool(
+                description="IMDb Scraper to get the IMDb data a given movie. "
+                "Given a movie name (in English), "
+                "finds the movie URL, rating, and scrapes the IMDb page (as Markdown)."
```
Comment (fjsj) [inline example/movies/ai_assistants.py:130]: A better description to the IMDbScraper tool helps the AI to call it more.

File: example/movies/ai_assistants.py [hunk 7]
```diff
@@ -116,7 +148,10 @@ def get_movies_backlog(self) -> str:
 
     @method_tool
     def add_movie_to_backlog(self, movie_name: str, imdb_url: str, imdb_rating: float) -> str:
-        """Add a movie to user's backlog. Must pass the movie_name, imdb_url, and imdb_rating."""
+        """
+        Add a movie to user's backlog. Must pass the movie_name, imdb_url, and imdb_rating.
+        Set imdb_rating to 0.0 if not available.
```
Comment (fjsj) [inline example/movies/ai_assistants.py:153]: Facilitates adding not launched movies.

File: example/movies/ai_assistants.py (hunk 5 above)
Comment (pamella) [reply example/movies/ai_assistants.py:90]: > it was hallucinating URLs.

I noticed that too when I tried to use it for the tour guide example.

File: example/movies/ai_assistants.py (hunk 3 above)
Comment (pamella) [reply example/movies/ai_assistants.py:56]: Nice!

**Review by pamella — APPROVED** [review]
:rocket:

**Pull request merged**